pyro/infer/mcmc/hmc.py

```diff
@@ -366,7 +372,7 @@ def setup(self, warmup_steps, *args, **kwargs):
             self._r_shapes[name] = site_value.shape
             self._r_numels[name] = site_value.numel()
 
-        self._validate_trace(trace)
+        self._get_valid_prototype_trace(trace)
```

Comment:
Would suggest a minor refactoring as follows:
 - Make the `_get_valid_prototype_trace` not take in any args, and just return a valid trace, or throw a `ValueError`
 - Move the call to this method early in setup, and use this trace for any subsequent setup. That way, we can terminate early if we don't find a good value for the trace.

   ```python
   def setup(..):
       trace = self._get_valid_prototype_trace()
       # subsequent operations
       ...
   ```

Reply:
In the setup, we need to know if we have `_has_enumerable_sites` to feed into `trace_prob_evaluator` . So I think that we need to understand the structure before finding a valid trace.

Reply:
I see - I think it might call for a slightly bigger reorganization. Let us just rename the method to `_set_valid_prototype_trace` to clearly designate that it is setting an attribute.